README.md
# GenSVM Python Package

[![Build Status](https://travis-ci.org/GjjvdBurg/PyGenSVM.svg?branch=master)](https://travis-ci.org/GjjvdBurg/PyGenSVM)
[![Documentation Status](https://readthedocs.org/projects/gensvm/badge/?version=latest)](https://gensvm.readthedocs.io/en/latest/?badge=latest)

This is the Python package for the GenSVM multiclass classifier by [Gerrit 
J.J. van den Burg](https://gertjanvandenburg.com) and [Patrick J.F. 
Groenen](https://personal.eur.nl/groenen/).

**Useful links:**

- [PyGenSVM on GitHub](https://github.com/GjjvdBurg/PyGenSVM)
- [PyGenSVM on PyPI](https://pypi.org/project/gensvm/)
- [Package documentation](https://gensvm.readthedocs.io/en/latest/)
- Journal paper: [GenSVM: A Generalized Multiclass Support Vector 
  Machine](http://www.jmlr.org/papers/v17/14-526.html) JMLR, 17(225):1−42, 
  2016.
- There is also an [R package](https://github.com/GjjvdBurg/RGenSVM)
- Or you can directly use [the C library](https://github.com/GjjvdBurg/GenSVM)


## Installation

**Before** GenSVM can be installed, a working NumPy installation is required. 
so GenSVM can be installed using the following command:

```bash
$ pip install numpy && pip install gensvm
```

If you encounter any errors, please [open an issue on 
GitHub](https://github.com/GjjvdBurg/PyGenSVM). Don't hesitate, you're helping 
to make this project better!


## Citing

If you use this package in your research please cite the paper, for instance 
using the following BibTeX entry::

```bib
@article{JMLR:v17:14-526,
        author  = {{van den Burg}, G. J. J. and Groenen, P. J. F.},
        title   = {{GenSVM}: A Generalized Multiclass Support Vector Machine},
        journal = {Journal of Machine Learning Research},
        year    = {2016},
        volume  = {17},
        number  = {225},
        pages   = {1-42},
        url     = {http://jmlr.org/papers/v17/14-526.html}
}
```

## Usage

The package contains two classes to fit the GenSVM model: [GenSVM] and 
[GenSVMGridSearchCV].  These classes respectively fit a single GenSVM model or 
fit a series of models for a parameter grid search. The interface to these 
classes is the same as that of classifiers in [Scikit-Learn]  so users 
familiar with Scikit-Learn should have no trouble using this package.  Below 
we will show some examples of using the GenSVM classifier and the 
GenSVMGridSearchCV class in practice.

In the examples we assume that we have loaded the [iris 
dataset](http://scikit-learn.org/stable/auto_examples/datasets/plot_iris_dataset.html) 
from Scikit-Learn as follows:


```python
>>> from sklearn.datasets import load_iris
>>> from sklearn.model_selection import train_test_split
>>> from sklearn.preprocessing import MaxAbsScaler
>>> X, y = load_iris(return_X_y=True)
>>> X_train, X_test, y_train, y_test = train_test_split(X, y)
>>> scaler = MaxAbsScaler().fit(X_train)
>>> X_train, X_test = scaler.transform(X_train), scaler.transform(X_test)
```

Note that we scale the data using the 
[MaxAbsScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MaxAbsScaler.html)
function. This scales the columns of the data matrix to ``[-1, 1]`` without 
breaking sparsity. Scaling the dataset can have a significant effect on the 
computation time of GenSVM and is [generally recommended for 
SVMs](https://stats.stackexchange.com/q/65094).


### Example 1: Fitting a single GenSVM model

Let's start by fitting the most basic GenSVM model on the training data:


```python
>>> from gensvm import GenSVM
>>> clf = GenSVM()
>>> clf.fit(X_train, y_train)
GenSVM(coef=0.0, degree=2.0, epsilon=1e-06, gamma='auto', kappa=0.0,
kernel='linear', kernel_eigen_cutoff=1e-08, lmd=1e-05,
max_iter=100000000.0, p=1.0, random_state=None, verbose=0,
weights='unit')
```

With the model fitted, we can predict the test dataset:

```python
>>> y_pred = clf.predict(X_test)
```

Next, we can compute a score for the predictions. The GenSVM class has a 
``score`` method which computes the 
[accuracy_score](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html)
for the predictions. In the GenSVM paper, the [adjusted Rand 
index](https://en.wikipedia.org/wiki/Rand_index#Adjusted_Rand_index) is often 
used to compare performance. We illustrate both options below (your results 
may be different depending on the exact train/test split):

```python
>>> clf.score(X_test, y_test)
1.0
>>> from sklearn.metrics import adjusted_rand_score
>>> adjusted_rand_score(clf.predict(X_test), y_test)
1.0
```

We can try this again by changing the model parameters, for instance we can 
turn on verbosity and use the Euclidean norm in the GenSVM model by setting ``p = 2``:

```python
>>> clf2 = GenSVM(verbose=True, p=2)
>>> clf2.fit(X_train, y_train)
Starting main loop.
Dataset:
    n = 112
    m = 4
    K = 3
Parameters:
    kappa = 0.000000
    p = 2.000000
    lambda = 0.0000100000000000
    epsilon = 1e-06

iter = 0, L = 3.4499531579689533, Lbar = 7.3369415851139745, reldiff = 1.1266786095824437
...
Optimization finished, iter = 4046, loss = 0.0230726364692517, rel. diff. = 0.0000009998645783
Number of support vectors: 9
GenSVM(coef=0.0, degree=2.0, epsilon=1e-06, gamma='auto', kappa=0.0,
    kernel='linear', kernel_eigen_cutoff=1e-08, lmd=1e-05,
    max_iter=100000000.0, p=2, random_state=None, verbose=True,
    weights='unit')
```

For other parameters that can be tuned in the GenSVM model, see [GenSVM].

### Example 2: Fitting a GenSVM model with a "warm start"

One of the key features of the GenSVM classifier is that training can be 
accelerated by using so-called "warm-starts". This way the optimization can be 
started in a location that is closer to the final solution than a random 
starting position would be. To support this, the ``fit`` method of the GenSVM 
class has an optional ``seed_V`` parameter. We'll illustrate how this can be 
used below.

We start with relatively large value for the ``epsilon`` parameter in the 
model. This is the stopping parameter that determines how long the 
optimization continues (and therefore how exact the fit is).

```python
>>> clf1 = GenSVM(epsilon=1e-3)
>>> clf1.fit(X_train, y_train)
...
>>> clf1.n_iter_
163
```

The ``n_iter_`` attribute tells us how many iterations the model did. Now, we 
can use the solution of this model to start the training for the next model:

```python
>>> clf2 = GenSVM(epsilon=1e-8)
>>> clf2.fit(X_train, y_train, seed_V=clf1.combined_coef_)
...
>>> clf2.n_iter_
3196
```

Compare this to a model with the same stopping parameter, but without the warm 
start:

```python
>>> clf2.fit(X_train, y_train)
...
>>> clf2.n_iter_
3699
```

So we saved about 500 iterations! This effect will be especially significant 
with large datasets and when you try out many parameter configurations. 
Therefore this technique is built into the [GenSVMGridSearchCV] class that can 
be used to do a grid search of parameters.

### Example 3: Running a GenSVM grid search

Often when we're fitting a machine learning model such as GenSVM, we have to 
try several parameter configurations to figure out which one performs best on 
our given dataset. This is usually combined with [cross 
validation](http://scikit-learn.org/stable/modules/cross_validation.html) to 
avoid overfitting. To do this efficiently and to make use of warm starts, the 
[GenSVMGridSearchCV] class is available. This class works in the same way as 
the 
[GridSearchCV](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)
class of [Scikit-Learn], but uses the GenSVM C library for speed.

To do a grid search, we first have to define the parameters that we want to 
vary and what values we want to try:

```python
>>> from gensvm import GenSVMGridSearchCV
>>> param_grid = {'p': [1.0, 2.0], 'lmd': [1e-8, 1e-6, 1e-4, 1e-2, 1.0], 'kappa': [-0.9, 0.0] }
```

For the values that are not varied in the parameter grid, the default values 
will be used. This means that if you want to change a specific value (such as 
``epsilon`` for instance), you can add this to the parameter grid as a 
parameter with a single value to try (e.g. ``'epsilon': [1e-8]``).

Running the grid search is now straightforward:

```python
>>> gg = GenSVMGridSearchCV(param_grid)
>>> gg.fit(X_train, y_train)
GenSVMGridSearchCV(cv=None, iid=True,
      param_grid={'p': [1.0, 2.0], 'lmd': [1e-06, 0.0001, 0.01, 1.0], 'kappa': [-0.9, 0.0]},
      refit=True, return_train_score=True, scoring=None, verbose=0)
```

Note that if we have set ``refit=True`` (the default), then we can use the 
[GenSVMGridSearchCV] instance to predict or score using the best estimator 
found in the grid search:

```python
>>> y_pred = gg.predict(X_test)
>>> gg.score(X_test, y_test)
1.0
```

A nice feature borrowed from `Scikit-Learn`_ is that the results from the grid 
search can be represented as a ``pandas`` DataFrame:

```python
>>> from pandas import DataFrame
>>> df = DataFrame(gg.cv_results_)
```

This can make it easier to explore the results of the grid search.

## Known Limitations

The following are known limitations that are on the roadmap for a future 
release of the package. If you need any of these features, please vote on them 
on the linked GitHub issues (this can make us add them sooner!).

1. [Support for sparse 
   matrices](https://github.com/GjjvdBurg/PyGenSVM/issues/1). NumPy supports 
   sparse matrices, as does the GenSVM C library. Getting them to work 
   together requires some additional effort. In the meantime, if you really 
   want to use sparse data with GenSVM (this can lead to significant 
   speedups!), check out the GenSVM C library.
2. [Specification of class misclassification 
   weights](https://github.com/GjjvdBurg/PyGenSVM/issues/3). Currently, 
   incorrectly classification an object from class A to class C is as bad as 
   incorrectly classifying an object from class B to class C. Depending on the 
   application, this may not be the desired effect. Adding class 
   misclassification weights can solve this issue.


## Questions and Issues

If you have any questions or encounter any issues with using this package, 
please ask them on [GitHub](https://github.com/GjjvdBurg/PyGenSVM).

## License

This package is licensed under the GNU General Public License version 3. 

Copyright (c) G.J.J. van den Burg, excluding the sections of the code that are 
explicitly marked to come from Scikit-Learn.

[Scikit-Learn]: http://scikit-learn.org/stable/index.html

[GenSVM]: https://gensvm.readthedocs.io/en/latest/#gensvm

[GenSVMGridSearchCV]: https://gensvm.readthedocs.io/en/latest/#gensvmgridsearchcv


CHANGELOG.md
## Change Log

### Version 0.2.7

- Deal with various deprecated features of sklearn

### Version 0.2.6

- Minor fixes to fix description on PyPI

### Version 0.2.5

- Build platform wheels for Linux and MacOS
- Minor improvements to the package

### Version 0.2.4

- Add support for retrieving support vectors

### Version 0.2.3

- Bugfix for prediction with gamma = 'auto'

### Version 0.2.2

- Add error when unsupported ShuffleSplits are used

### Version 0.2.1

- Update docs
- Speed up unit tests

### Version 0.2.0

- Add support for interrupting training and retreiving partial results
- Allow specification of sample weights in GenSVM.fit()
- Extract per-split durations from grid search results
- Add pre-defined parameter grids 'tiny', 'small', and 'full'
- Add code for prediction with kernels
- Add unit tests
- Change default coef in poly kernel to 1.0 for inhomogeneous kernel
- Minor bugfixes, documentation improvement, and code cleanup
- Add continuous integration through Travis-CI.

### Version 0.1.6

- Fix segfault caused by error function in C library.
- Add "load_default_grid" function to gensvm.gridsearch


gensvm/models.py
# -*- coding: utf-8 -*-

"""
"""

from __future__ import print_function, division

import numpy as np
import warnings

from sklearn.base import BaseEstimator
from sklearn.exceptions import ConvergenceWarning, FitFailedWarning
from sklearn.preprocessing import LabelEncoder
from sklearn.utils import check_X_y, check_random_state
from sklearn.utils.multiclass import type_of_target
from sklearn.utils.validation import check_is_fitted

from . import pyx_gensvm


def _fit_gensvm(X, y, p, lmd, kappa, epsilon, weight_idx, kernel, gamma, coef, 
        degree, kernel_eigen_cutoff, verbose, max_iter, random_state=None):

    # process the random state
    rnd = check_random_state(random_state)

    # set the verbosity in GenSVM
    pyx_gensvm.set_verbosity_wrap(verbose)

    # run the actual training
    raw_coef_, n_SV_, n_iter_, training_error_, status_ = pyx_gensvm.train_wrap(
            X, y, p, lmd, kappa, epsilon, weight_idx, kernel, gamma, coef, 
            degree, kernel_eigen_cutoff, max_iter, 
            rnd.randint(np.iinfo('i').max))

    # process output
    if status_ == 1 and verbose > 0:
        warnings.warn("GenSVM optimization prematurely ended due to a "
                "incorrect step in the optimization algorithm.", 
                FitFailedWarning)

    if status_ == 2 and verbose > 0:
        warnings.warn("GenSVM failed to converge, increase "
                "the number of iterations.", ConvergenceWarning)

    coef_ = raw_coef_[1:, :]
    intercept_ = raw_coef_[0, :]

    return coef_, intercept_, n_iter_, n_SV_


class GenSVM(BaseEstimator):
    """Generalized Multiclass Support Vector Machine Classification.

    This class implements the basic GenSVM classifier. GenSVM is a generalized 
    multiclass SVM which is flexible in the weighting of misclassification 
    errors. It is this flexibility that makes it perform well on diverse 
    datasets.

    This methods of this class use the GenSVM C library for the actual 
    computations.

    Parameters
    ----------
    p : float, optional (default=1.0)
        Parameter for the L_p norm of the loss function (1.0 <= p <= 2.0)

    lmd : float, optional (default=1e-5)
        Parameter for the regularization term of the loss function (lmd > 0)

    kappa : float, optional (default=0.0)
        Parameter for the hinge function in the loss function (kappa > -1.0)

    weight_idx : int, optional (default=1)
        Type of sample weights to use (1 = unit weights, 2 = size correction 
        weights)

    kernel : string, optional (default='linear')
        Specify the kernel type to use in the classifier. It must be one of 
        'linear', 'poly', 'rbf', or 'sigmoid'.

    gamma : float, optional (default=1.0)
        Kernel parameter for the rbf, poly, and sigmoid kernel

    coef : float, optional (default=0.0)
        Kernel parameter for the poly and sigmoid kernel

    degree : float, optional (default=2.0)
        Kernel parameter for the poly kernel

    kernel_eigen_cutoff : float, optional (default=1e-8)
        Cutoff point for the reduced eigendecomposition used with 
        kernel-GenSVM. Eigenvectors for which the ratio between their 
        corresponding eigenvalue and the largest eigenvalue is smaller than the 
        cutoff will be dropped.

    verbose : int, (default=0)
        Enable verbose output

    max_iter : int, (default=1e8)
        The maximum number of iterations to be run.


    Attributes
    ----------
    coef_ : array, shape = [n_features, n_classes-1]
        Weights assigned to the features (coefficients in the primal problem)

    intercept_ : array, shape = [n_classes]
        Constants in the decision function

    n_iter_ : int
        The number of iterations that were run during training.

    n_support_ : int
        The number of support vectors that were found


    References
    ----------
    * Van den Burg, G.J.J. and Groenen, P.J.F.. GenSVM: A Generalized 
    Multiclass Support Vector Machine. Journal of Machine Learning Research, 
    17(225):1--42, 2016.

    """

    def __init__(self, p=1.0, lmd=1e-5, kappa=0.0, epsilon=1e-6, weight_idx=1, 
            kernel='linear', gamma=1.0, coef=0.0, degree=2.0, 
            kernel_eigen_cutoff=1e-8, verbose=0, random_state=None, 
            max_iter=1e8):
        self.p = p
        self.lmd = lmd
        self.kappa = kappa
        self.epsilon = epsilon
        self.weight_idx = weight_idx
        self.kernel = kernel
        self.gamma = gamma
        self.coef = coef
        self.degree = degree
        self.kernel_eigen_cutoff = kernel_eigen_cutoff
        self.verbose = verbose
        self.random_state = random_state
        self.max_iter = max_iter


    def fit(self, X, y):
        if not 1.0 <= self.p <= 2.0:
            raise ValueError("Value for p should be within [1, 2]; got p = %r)" 
                    % self.p)
        if not self.kappa > -1.0:
            raise ValueError("Value for kappa should be larger than -1; got "
                    "kappa = %r" % self.kappa)
        if not self.lmd > 0:
            raise ValueError("Value for lmd should be larger than 0; got "
                    "lmd = %r" % self.lmd)
        if not self.epsilon > 0:
            raise ValueError("Value for epsilon should be larger than 0; got "
                    "epsilon = %r" % self.epsilon)
        X, y_org = check_X_y(X, y, accept_sparse=False, dtype=np.float64, 
                order="C")

        y_type = type_of_target(y_org)
        if y_type not in ["binary", "multiclass"]:
            raise ValueError("Label type not allowed for GenSVM: %r" % y_type)

        # This is necessary because GenSVM expects classes to go from 1 to 
        # n_class
        self.encoder = LabelEncoder()
        y = self.encoder.fit_transform(y_org)
        y += 1

        self.coef_, self.intercept_, self.n_iter_, self.n_support_ = \
                _fit_gensvm(X, y, self.p, self.lmd, self.kappa, self.epsilon, 
                        self.weight_idx, self.kernel, self.gamma, self.coef, 
                        self.degree, self.kernel_eigen_cutoff, self.verbose, 
                        self.max_iter, self.random_state)
        return self


    def predict(self, X):
        check_is_fitted(self, "coef_")

        V = np.vstack((self.intercept_, self.coef_))
        predictions = pyx_gensvm.predict_wrap(X, V)

        # Transform the classes back to the original form
        predictions -= 1
        outcome = self.encoder.inverse_transform(predictions)

        return outcome
